rofl-scheduler/src/proxy/domain.rs

```diff
@@ -0,0 +1,350 @@
+        logger: slog::Logger,
+    ) {
+        let mut tasks = JoinSet::new();
+        let (retry_tx, retry_rx) = mpsc::unbounded_channel();
+
+        // Spawn retry scheduler.
+        tasks.spawn(Self::run_retry_scheduler(
+            retry_rx,
+            work_tx.clone(),
+            logger.clone(),
+        ));
+
+        // Prepare DNS resolver.
+        let mut name_servers = NameServerConfigGroup::google_https();
+        name_servers.merge(NameServerConfigGroup::cloudflare_https());
+        name_servers.merge(NameServerConfigGroup::quad9_https());
+
+        let mut builder = hickory_resolver::Resolver::builder_with_config(
+            hickory_resolver::config::ResolverConfig::from_parts(None, vec![], name_servers),
+            hickory_resolver::name_server::TokioConnectionProvider::default(),
+        );
+        builder.options_mut().validate = true; // Enable DNSSEC validation.
+        let resolver = builder.build();
+
+        // Spawn domain verification workers.
+        for _ in 0..workers {
+            tasks.spawn(Self::run_worker(
+                work_rx.clone(),
+                retry_tx.clone(),
+                notifier.clone(),
+                resolver.clone(),
+                logger.clone(),
+            ));
+        }
+
+        tasks.join_all().await;
+    }
+
+    /// Run the retry scheduler that manages failed verifications.
+    ///
+    /// The scheduler maintains a list of pending retries and sleeps until the next
+    /// retry is due. When a verification is ready, it's sent back to the worker queue
+    /// for immediate processing.
+    async fn run_retry_scheduler(
+        mut retry_rx: mpsc::UnboundedReceiver<DomainVerification>,
+        work_tx: tokio_mpmc::Sender<DomainVerification>,
+        logger: slog::Logger,
+    ) -> Result<()> {
+        let mut pending_retries = BinaryHeap::new();
+
+        loop {
+            tokio::select! {
+                // Handle new retry requests.
+                verification = retry_rx.recv() => {
+                    match verification {
+                        Some(mut verification) => {
+                            if verification.is_cancelled() || !verification.schedule_retry() {
+                                slog::warn!(logger, "giving up on domain verification";
+                                    "domain" => &verification.domain,
+                                    "instance_id" => ?verification.instance_id,
+                                );
+                            } else {
+                                slog::debug!(logger, "scheduling domain verification retry";
+                                    "domain" => &verification.domain,
+                                    "instance_id" => ?verification.instance_id,
+                                    "retry_at" => ?verification.retry_at,
+                                );
+                                pending_retries.push(verification);
```

Comment:
```suggestion
                                pending_retries.push(Reverse(verification));
```
I think that by default this is a max-heap.

Reply:
But the `cmp` is already reversed so it should be fine (e.g. items with lowest `retry_at` will be considered "greater than")?
```rust
impl Ord for DomainVerification {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other.retry_at.cmp(&self.retry_at)
    }
}
```

Reply:
Added a test in #2330 to make sure this is and remains correct.